### ISSUE TITLE
Task Fusion

### ISSUE THREAD

**Issue by mrocklin** [ContinuumIO/dask#18]
We want an optimization pass that fuses linear sequences of tasks.  Note the disappearance of 'w' and 'y'

```Python
>>> d = {'v': (inc, 'y'),
...      'u': (inc, 'w'),
...      'w': (inc, 'x'), 
...      'x': (inc, 'y'), 
...      'y': (inc, 'z'), 
...      'z': (add, 'a', 'b'), 
...      'a': (inc, 'c'), 'b': (inc, 'd'),
...      'c': 1, 'd': 2}
>>> fuse(d)
{'u': (inc, (inc, 'x')),
 'x': (inc, 'y'),
 'v': (inc, 'y'),
 'y': ((inc, (add, 'a', 'b'))), 
 'a': (inc, 'c'), 'b': (inc, 'd'),
 'c': 1, 'd': 2}
```

We only want to do this when the task has one real dependency.  We don't want to destroy potential parallelism.

@eriknw this seems up your alley.  From my perspective this just became relatively high priority.  I might take a crack in the next few days so do let me know if you take it on.

**Comment by eriknw** [comment]
I'm on it.

**Comment by eriknw** [comment]
@mrocklin, what are your thoughts on constant propagation, which is closely related to `fuse` and could potentially be part of `fuse`?  For example, `z` could be replaced with `'z': (add, 1, 2)`, which would allow it to be fused into `w`.  I think constant propagation should be a separate function with a `filter` keyword that allows the user to determine what constants to fold.

**Comment by mrocklin** [comment]
Agreed on all counts.

The one thing against working to inline constants is that I don't yet have a practical motivating use case.

**Comment by eriknw** [comment]
The use case is more aggressive task fusion.  An argument against it is that data transfer may be expensive, in which case it's best if the scheduler/executor manages input data.

**Comment by mrocklin** [comment]
I guess argument for is 

1.  It saves us a dereference
2.  It allows us things like `fuse` to work more aggressively (the simple example above is a good example of this)

The workflows we're currently sending against dask don't motivate 1.  Generally our tasks are large.  This would be more important in a lower level compiler.

Number 2 is more motivating though.  It might also make sense to look at other optimizations passes commonly performed by compilers and see which of them are and are not useful for the workflows we're likely to handle. 

BTW, I factored the threaded scheduler/get to a general async one.  I'm now processing bags with `multiprocessing.Pool.apply_async`.  Still no attempt at data locality (the hard part) which is why optimizations like `fuse` become really important, they fuse the parts that should stay local so that the scheduler can't mismanage them.

**Comment by eriknw** [comment]
Is this the desired behavior?  Note that two tasks now depend on `'y'`:
```python
>>> d = {'NEW': (inc, 'y'),
...      'w': (inc, 'x'),
...      'x': (inc, 'y'),
...      'y': (inc, 'z'),
...      'z': (add, 'a', 'b'),
...      'a': 1, 'b': 2}

>>> dask.core.fuse(d)
{'NEW': (inc, 'y'),
 'a': 1,
 'b': 2,
 'w': (inc, (inc, 'y')),
 'y': (inc, 'z'),
 'z': (add, 'a', 'b')}
```

**Comment by mrocklin** [comment]
Switching to images (btw, see `dask.dot.dot_graph`)

<img src="https://cloud.githubusercontent.com/assets/306380/5989369/76a6648c-a93a-11e4-9140-f29fa881ef2e.png" align=left>
<img src="https://cloud.githubusercontent.com/assets/306380/5989371/7a921b40-a93a-11e4-9255-2822d39cd3d8.png" align=left>

The behavior around `w,x,y` is good and desired behavior (kicking myself for not thinking about this.)

Looking at these images I now think we may also want to collapse out the storage of `z`.  I was wrong in my original example.  We're not destroying parallelism by removing it.  We wouldn't want to go any further though.

**Comment by mrocklin** [comment]
Wow, github really doesn't like html in comments.  Those turned out really funky when I tried to put them side by side.

**Comment by eriknw** [comment]
Great.  So the answer is yes... with the caveat that we may want to fuse `z` into `y` as well.  I was about to ask about that as well.  I'll have something to share soon.

Shouldn't the second dot graph show two `inc` circles between `w` and `y`?

**Comment by mrocklin** [comment]
OK, I've updated the header comment with, I think, a correct example.  Reproducing here with side-by-side image

![fusion](https://cloud.githubusercontent.com/assets/306380/5989432/00e58072-a93d-11e4-95f0-eb01be1cc478.png)

Yeah, it would be nice to have better DOT printing with nested operations.

**Comment by mrocklin** [comment]
I'm excited for `fuse` btw.  This should give us a decently powerful system with relatively little work. 

One way to think of it is that anything that `toolz` could previously do in a streaming way we can probably now do in a parallel way.

**Comment by eriknw** [comment]
Wait, so you don't want to fuse *any* constants?  In the new example, why shouldn't `c` fuse with `a` and `d` fuse with `b`?

**Comment by mrocklin** [comment]
The two `inc`s at the bottom of the graph could be expensive and can run in parallel.  They should remain.

**Comment by mrocklin** [comment]
The pattern we want to destroy is a long vertical chain.  Everything else we probably want to keep.

**Comment by eriknw** [comment]
Sure... the constants `c` and `d` are part of long vertical chains, which will run in parallel via `a` and `b`.  Fusing the constants in this case does not change that.

**Comment by mrocklin** [comment]
Ah, if we're talking about inlining constants then I like your previous statement that it's a good idea but probably a separate optimization.

I might not understand what you're suggesting though.  What dict would you expect as output from the example above?

**Comment by eriknw** [comment]
I'm currently writing tests, and I'll share the expected dict shortly.

`c` and `d` would be fused if they weren't constants, because they are used only once.  Do you want to exclude them because they are constants?

**Comment by mrocklin** [comment]
OK, sorry, I see what you're saying now.

My expectation was that they would be untouched.  But now I can't actually provide a principled reason as to why that should be.  So I say go ahead with not treating them specially.

**Comment by eriknw** [comment]
Okay, so when fusing the constants from the given input (same as in header comment):
```python
d = {
    'v': (inc, 'y'),
    'u': (inc, 'w'),
    'w': (inc, 'x'),
    'x': (inc, 'y'),
    'y': (inc, 'z'),
    'z': (add, 'a', 'b'),
    'a': (inc, 'c'),
    'b': (inc, 'd'),
    'c': 1,
    'd': 2,
}
```
the resulting dict you shared is:
```python
orig = {
    'u': (inc, (inc, 'x')),
    'x': (inc, 'y'),
    'v': (inc, 'y'),
    'y': (inc, (add, 'a', 'b')),
    'a': (inc, 1),
    'b': (inc, 2),
}
```
I think the answer should be:
```python
new = {
    'u': (inc, (inc, (inc, 'y'))),
    'v': (inc, 'y'),
    'y': (inc, (add, 'a', 'b')),
    'a': (inc, 1),
    'b': (inc, 2),
}
```
Let's see if I can figure out out to show images...
# Orig:
![orig](https://cloud.githubusercontent.com/assets/2058401/5989746/c96e73a0-a956-11e4-82e1-320cdbb77ae8.png)

# New:
![new](https://cloud.githubusercontent.com/assets/2058401/5989745/c95d67d6-a956-11e4-912e-dc0cd434f108.png)

**Comment by mrocklin** [comment]
Closing this for now.  There are probably improvements that can be made but, for the most part, this works well.
